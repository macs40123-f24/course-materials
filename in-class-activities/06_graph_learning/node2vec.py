# Adapted from: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/node2vec.py
import matplotlib.pyplot as plt
import torch
from sklearn.manifold import TSNE
from torch_geometric.datasets import Planetoid
from torch_geometric.nn import Node2Vec

def train():
    model.train()
    total_loss = 0
    for pos_rw, neg_rw in loader:
        optimizer.zero_grad()
        loss = model.loss(pos_rw.to(device), neg_rw.to(device))
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    return total_loss / len(loader)


@torch.no_grad()
def test():
    model.eval()
    z = model()
    acc = model.test(
        train_z=z[data.train_mask],
        train_y=data.y[data.train_mask],
        test_z=z[data.test_mask],
        test_y=data.y[data.test_mask],
        max_iter=150,
    )
    return acc


@torch.no_grad()
def plot_points(colors):
    model.eval()
    z = model().cpu().numpy()
    z = TSNE(n_components=2).fit_transform(z)
    y = data.y.cpu().numpy()

    plt.figure(figsize=(8, 8))
    for i in range(dataset.num_classes):
        plt.scatter(z[y == i, 0], z[y == i, 1], s=20, color=colors[i])
    plt.axis('off')
    plt.savefig('node2vec.png')


if __name__ == '__main__':
    # Cora Citation Network Data: https://paperswithcode.com/dataset/cora
    # nodes = publications, edges = citations
    dataset = Planetoid('/project/macs40123/', name='Cora')
    data = dataset[0]

    # Train on GPU if available; takes <=1 min on GPU, ~15 min on CPU
    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    
    # Initialize Node2Vec model with hyperparameters
    model = Node2Vec(
        data.edge_index,
        embedding_dim=128,
        walk_length=20,
        context_size=10,
        walks_per_node=10,
        num_negative_samples=1,
        p=1.0,
        q=1.0,
        sparse=True,
    ).to(device)
    
    # define data loader (generates batches of rand walks) and optimizer
    # for training Node2Vec model:
    loader = model.loader(batch_size=128, shuffle=True)
    optimizer = torch.optim.SparseAdam(list(model.parameters()), lr=0.01)

    # Train for 100 epochs and print progress
    for epoch in range(1, 101):
        loss = train()
        acc = test()
        print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}, Acc: {acc:.4f}')

    # Plot embeddings in 2D, with different colors for 7 classes of articles/nodes
    # Did node2vec learn to distinguish the node classes based on network structure?
    colors = [
        '#ffc0cb', '#bada55', '#008080', '#420420', '#7fe5f0', '#065535', '#ffd700'
    ]
    plot_points(colors)
