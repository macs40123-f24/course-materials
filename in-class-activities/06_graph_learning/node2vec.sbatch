#!/bin/bash
#SBATCH --job-name=node2vec
#SBATCH --output=node2vec.out
#SBATCH --error=node2vec.err
#SBATCH --time=00:30:00
#SBATCH --nodes=1
#SBATCH --partition=gpu
#SBATCH --ntasks=1
#SBATCH --gres=gpu:1
#SBATCH --account=macs40123
#SBATCH --mem-per-cpu=30G

module load python cuda

python node2vec.py
