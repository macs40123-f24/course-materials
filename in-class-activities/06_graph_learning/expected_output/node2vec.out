Epoch: 001, Loss: 8.1181, Acc: 0.1590
Epoch: 002, Loss: 6.0729, Acc: 0.1840
Epoch: 003, Loss: 4.9560, Acc: 0.2000
Epoch: 004, Loss: 4.1413, Acc: 0.2330
Epoch: 005, Loss: 3.4914, Acc: 0.2690
Epoch: 006, Loss: 2.9711, Acc: 0.2870
Epoch: 007, Loss: 2.5617, Acc: 0.3300
Epoch: 008, Loss: 2.2307, Acc: 0.3520
Epoch: 009, Loss: 1.9642, Acc: 0.3770
Epoch: 010, Loss: 1.7479, Acc: 0.4200
Epoch: 011, Loss: 1.5774, Acc: 0.4450
Epoch: 012, Loss: 1.4366, Acc: 0.4760
Epoch: 013, Loss: 1.3261, Acc: 0.4970
Epoch: 014, Loss: 1.2371, Acc: 0.5170
Epoch: 015, Loss: 1.1641, Acc: 0.5370
Epoch: 016, Loss: 1.1069, Acc: 0.5580
Epoch: 017, Loss: 1.0621, Acc: 0.5760
Epoch: 018, Loss: 1.0267, Acc: 0.5920
Epoch: 019, Loss: 0.9976, Acc: 0.6130
Epoch: 020, Loss: 0.9743, Acc: 0.6270
Epoch: 021, Loss: 0.9542, Acc: 0.6380
Epoch: 022, Loss: 0.9391, Acc: 0.6400
Epoch: 023, Loss: 0.9245, Acc: 0.6520
Epoch: 024, Loss: 0.9155, Acc: 0.6560
Epoch: 025, Loss: 0.9046, Acc: 0.6500
Epoch: 026, Loss: 0.8945, Acc: 0.6600
Epoch: 027, Loss: 0.8881, Acc: 0.6640
Epoch: 028, Loss: 0.8807, Acc: 0.6700
Epoch: 029, Loss: 0.8773, Acc: 0.6780
Epoch: 030, Loss: 0.8719, Acc: 0.6920
Epoch: 031, Loss: 0.8676, Acc: 0.6910
Epoch: 032, Loss: 0.8631, Acc: 0.6980
Epoch: 033, Loss: 0.8594, Acc: 0.7020
Epoch: 034, Loss: 0.8578, Acc: 0.7010
Epoch: 035, Loss: 0.8545, Acc: 0.7050
Epoch: 036, Loss: 0.8523, Acc: 0.7040
Epoch: 037, Loss: 0.8516, Acc: 0.7090
Epoch: 038, Loss: 0.8488, Acc: 0.7110
Epoch: 039, Loss: 0.8465, Acc: 0.7060
Epoch: 040, Loss: 0.8438, Acc: 0.6990
Epoch: 041, Loss: 0.8435, Acc: 0.7110
Epoch: 042, Loss: 0.8420, Acc: 0.7090
Epoch: 043, Loss: 0.8401, Acc: 0.7010
Epoch: 044, Loss: 0.8407, Acc: 0.7080
Epoch: 045, Loss: 0.8385, Acc: 0.7040
Epoch: 046, Loss: 0.8381, Acc: 0.7130
Epoch: 047, Loss: 0.8365, Acc: 0.7120
Epoch: 048, Loss: 0.8362, Acc: 0.7160
Epoch: 049, Loss: 0.8359, Acc: 0.7110
Epoch: 050, Loss: 0.8351, Acc: 0.7130
Epoch: 051, Loss: 0.8345, Acc: 0.7220
Epoch: 052, Loss: 0.8335, Acc: 0.7310
Epoch: 053, Loss: 0.8329, Acc: 0.7310
Epoch: 054, Loss: 0.8324, Acc: 0.7280
Epoch: 055, Loss: 0.8322, Acc: 0.7230
Epoch: 056, Loss: 0.8324, Acc: 0.7270
Epoch: 057, Loss: 0.8298, Acc: 0.7340
Epoch: 058, Loss: 0.8308, Acc: 0.7260
Epoch: 059, Loss: 0.8315, Acc: 0.7230
Epoch: 060, Loss: 0.8297, Acc: 0.7200
Epoch: 061, Loss: 0.8296, Acc: 0.7210
Epoch: 062, Loss: 0.8307, Acc: 0.7270
Epoch: 063, Loss: 0.8288, Acc: 0.7310
Epoch: 064, Loss: 0.8295, Acc: 0.7340
Epoch: 065, Loss: 0.8280, Acc: 0.7310
Epoch: 066, Loss: 0.8281, Acc: 0.7300
Epoch: 067, Loss: 0.8280, Acc: 0.7270
Epoch: 068, Loss: 0.8284, Acc: 0.7340
Epoch: 069, Loss: 0.8273, Acc: 0.7300
Epoch: 070, Loss: 0.8294, Acc: 0.7360
Epoch: 071, Loss: 0.8277, Acc: 0.7310
Epoch: 072, Loss: 0.8272, Acc: 0.7260
Epoch: 073, Loss: 0.8267, Acc: 0.7180
Epoch: 074, Loss: 0.8268, Acc: 0.7210
Epoch: 075, Loss: 0.8274, Acc: 0.7300
Epoch: 076, Loss: 0.8268, Acc: 0.7240
Epoch: 077, Loss: 0.8260, Acc: 0.7200
Epoch: 078, Loss: 0.8259, Acc: 0.7240
Epoch: 079, Loss: 0.8274, Acc: 0.7250
Epoch: 080, Loss: 0.8259, Acc: 0.7270
Epoch: 081, Loss: 0.8253, Acc: 0.7270
Epoch: 082, Loss: 0.8255, Acc: 0.7280
Epoch: 083, Loss: 0.8256, Acc: 0.7210
Epoch: 084, Loss: 0.8260, Acc: 0.7140
Epoch: 085, Loss: 0.8259, Acc: 0.7150
Epoch: 086, Loss: 0.8269, Acc: 0.7220
Epoch: 087, Loss: 0.8256, Acc: 0.7160
Epoch: 088, Loss: 0.8241, Acc: 0.7180
Epoch: 089, Loss: 0.8248, Acc: 0.7200
Epoch: 090, Loss: 0.8266, Acc: 0.7190
Epoch: 091, Loss: 0.8250, Acc: 0.7230
Epoch: 092, Loss: 0.8257, Acc: 0.7180
Epoch: 093, Loss: 0.8244, Acc: 0.7260
Epoch: 094, Loss: 0.8251, Acc: 0.7180
Epoch: 095, Loss: 0.8253, Acc: 0.7180
Epoch: 096, Loss: 0.8243, Acc: 0.7120
Epoch: 097, Loss: 0.8241, Acc: 0.7080
Epoch: 098, Loss: 0.8248, Acc: 0.7140
Epoch: 099, Loss: 0.8248, Acc: 0.7160
Epoch: 100, Loss: 0.8243, Acc: 0.7240
